src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs

```diff
@@ -66,23 +72,66 @@ internal HttpInMetricsListener(string name, Meter meter, AspNetCoreMetricsInstru
 
     public override void OnEventWritten(string name, object payload)
     {
-        if (name == OnStopEvent)
+        switch (name)
         {
-            if (this.emitOldAttributes)
-            {
-                this.OnEventWritten_Old(name, payload);
-            }
+            case OnUnhandledDiagnosticsExceptionEvent:
+            case OnUnhandledHostingExceptionEvent:
+                {
+                    if (this.emitNewAttributes)
+                    {
+                        this.OnExceptionEventWritten(name, payload);
+                    }
+                }
+
+                break;
+            case OnStopEvent:
+                {
+                    if (this.emitOldAttributes)
+                    {
+                        this.OnEventWritten_Old(name, payload);
+                    }
+
+                    if (this.emitNewAttributes)
+                    {
+                        this.OnEventWritten_New(name, payload);
+                    }
+                }
+
+                break;
+        }
+    }
 
-            if (this.emitNewAttributes)
-            {
-                this.OnEventWritten_New(name, payload);
-            }
+    public void OnExceptionEventWritten(string name, object payload)
+    {
+        // We need to use reflection here as the payload type is not a defined public type.
+        if (!TryFetchException(payload, out Exception exc) || !TryFetchHttpContext(payload, out HttpContext ctx))
+        {
+            AspNetCoreInstrumentationEventSource.Log.NullPayload(nameof(HttpInListener), nameof(this.OnExceptionEventWritten), HttpServerDurationMetricName);
+            return;
         }
+
+        ctx.Items.Add(ErrorTypeKey, exc.GetType().FullName);
```

Comment:
Would the framework create an activity even if only add AspNetCore metrics instrumentation? If yes, then I think we could add this information to `Activity.Current`. Do you know of any obvious pros and cons to adding this information to `HttpContext` vs `Activity.Current`?

Reply:
yes, the activity will always be there. Perf was the main reason to choose `HttpContext.Items`. However, we could use custom property as well on activity. Here are the results from benchmarking

Results when trace is also enabled and activity will have additional tags.

|                    Method |     Mean |    Error |   StdDev |   Median |   Gen0 | Allocated |
|-------------------------- |---------:|---------:|---------:|---------:|-------:|----------:|
|            UseActivityTag | 82.36 ns | 1.691 ns | 1.736 ns | 83.06 ns | 0.0085 |      40 B |
| UseActivityCustomProperty | 81.98 ns | 1.133 ns | 1.004 ns | 81.54 ns |      - |         - |
|           UseContextItems | 81.11 ns | 1.656 ns | 3.420 ns | 79.67 ns |      - |         - |


Results when only metrics will be enabled and activity will only have one tag

|                    Method |     Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------------------- |---------:|---------:|---------:|-------:|----------:|
|            UseActivityTag | 55.56 ns | 0.572 ns | 0.535 ns | 0.0085 |      40 B |
| UseActivityCustomProperty | 80.94 ns | 0.538 ns | 0.477 ns |      - |         - |
|           UseContextItems | 80.50 ns | 0.901 ns | 0.843 ns |      - |         - |

Based on above, I have updated the implementation to use custom property. We could follow the same thing on HttpClient side where we need to add `error.type`.

Reply:
IIRC the first time you add a custom property to activity you pay for an allocation. Are we already adding one somewhere else? Seems a bit odd that the allocation isn't showing in the benchmark.